ssl/statem/statem_lib.c

```diff
@@ -481,10 +481,9 @@ MSG_PROCESS_RETURN tls_process_cert_verify(SSL *s, PACKET *pkt)
     }
 #ifndef OPENSSL_NO_GOST
     {
-        int pktype = EVP_PKEY_id(pkey);
-        if (pktype == NID_id_GostR3410_2001
-            || pktype == NID_id_GostR3410_2012_256
-            || pktype == NID_id_GostR3410_2012_512) {
+        if (EVP_PKEY_is_a(pkey, "gost2001")
+            || EVP_PKEY_is_a(pkey, "gost2012_256")
+            || EVP_PKEY_is_a(pkey, "gost2012_512")) {
```

Comment:
This has me ponder if the names shouldn't include the OID, i.e. a gost provider should have something like this as an algorithm name: "gost-3410-2012-256:1.2.643.7.1.2.1.1.1.1"

(I chose "gost-3410-2012-256" from http://www.watersprings.org/pub/id/draft-deremin-rfc4491-bis-00.html#rfc.appendix.A)

Reply:
There can be various curves identified by various parameter OIDs but the key is still gost2012_256/gost2012_512 and they are processed the same way.

Reply:
Yes?  I was only thinking of what's the safest way to identify an external key type.  The OID is much more specific than a name that may or may not be "gost2012_256" exactly.  If someone else implements it, they might use names that exist in an RFC rather than the OpenSSL naming "standard".

This is part of figuring out how to play with the unknown with some amount of flexibility and weigh what we demand against that.  Since a dotted OID is very specific, and such a thing exist for this key type, demanding that provider authors include that as a name is probably easier than demanding that they following our naming standard.  At least from my personal perspective, I'd find it more natural.

Reply:
Well, I got your point now. I need to think more about it.

Reply:
I backed all of the EVP_PKEY_id() code anyway, so resolving this.